### ISSUE TITLE
[BUG]: State is lost when using the new idle/reconnect code

### ISSUE THREAD

**Issue by mlfarrell** [KittyCAD/modeling-app#2966]
### Describe the bug

Before, this worked fine.  So if we _have_ to reset the entire connection, at least lets not do it until we absolutely have to.  And if we do, we can't lose user settings.

### Steps to Reproduce

https://github.com/KittyCAD/modeling-app/assets/7664370/d46adfa7-d9ab-4504-bf47-83cbd73e99c3

### Expected Behavior

it not do that

### Desktop OS

_

### Version

_

**Comment by lf94** [comment]
(Haha @ issues trading :laughing: )

So yeah we lose state because you minimize the window and that fires a "visibilitychange" to "hidden". Unfortunately there's no way to differentiate between minimized, partially hidden, laptop closed, tab switched, etc. We will definitely need to record the state of before teardown and reconnect and send back what the user had done before since tab switching is a common pattern.

The alternative is we remove this type of idling. There's also timeout based idling as well that will have the same issue but maybe less of an issue since the user is actually away.

@jessfraz @Irev-Dev what do you want to do

**Comment by mlfarrell** [comment]
yea but there should be a timer/timeout?  it used to work fine for me on linux for a _long_ while without losing any connection.  this way its extremely jarring.

**Comment by lf94** [comment]
Ah yeah that'd be a way to temporarily fix this up - good idea :D

Before we had no idle logic so losing a connection was just a typical connection loss heh

**Comment by Irev-Dev** [comment]
Yeah I think we can pick a few bits of state that's jarring when lost, but don't need to go crazy, like exiting sketch I think is fine.
But grid should stay, and we can even grab the camera position and set it instead of the normal zoom to fit when reconnecting after idle.

**Comment by lf94** [comment]
I'll hammer out the camera position stuff but the grid setting just appears to not be saving at all

**Comment by Irev-Dev** [comment]
We decided against the camera shit, not worth it.

**Comment by mlfarrell** [comment]
sorry but this isn't fixed.  just happened to me again.  I barely was alt-tabbed away for 15-30 seconds (stream is open on my second monitor, clearly visible), and when I tabbed back, I got an ugly jarring black-white refresh.

**Comment by lf94** [comment]
@Irev-Dev may have closed it on accident, thinking it was the final solution to this

You're right that after 30 seconds of inactivity, it should teardown. It going black is actually undeterministic - I guess we can be better on that :) Thanks for re-opening so we can track this issue

In order for this issue to be properly resolved we need the engine to return camera state on some endpoint which shouldn't be too hard for me to go do

**Comment by mlfarrell** [comment]
I also wouldn't mind a power user setting to get it to just stop doing this, since this was never an issue for me in the first place.  aka, I never once had the stream drop out on me, even after leaving it for 5+ minutes at a time.  so it feels a little too intrusive to me

**Comment by mlfarrell** [comment]
don't mean to harp on this, but this is the sort of UX-irk that would cause me to close a product and literally never use it again.  so its worth bringing up.

**Comment by lf94** [comment]
The idling state behavior is less about fixing any issues and more about resource management but yes a setting to adjust the idle timeout would be nice for sure! And all good the app is for many types of people so addressing the annoyances is part of the job :)

**Comment by Irev-Dev** [comment]
Oh I didn't close this so much as I merged https://github.com/KittyCAD/modeling-app/pull/2970 which had the `closes` keyword in it, sorry didn't mean to.

Second monitor is interesting

**Comment by mlfarrell** [comment]
yea resource management side is tricky.  I think we should increase the idle timer at least a bit more, that'll remove most of the annoyances.  the other things may require me diving more into engine code (someday when I have the time) to see if I can make the reconnection less hitchy.

**Comment by Irev-Dev** [comment]
if we really can't detect a proper isn't visible anymore event, as might be the case with a second monitor, that might be a good argument to keep the camera position client side and snap to it on reconnection.

Potentially making a disconnect and reconnect seamless if we do it right.

**Comment by mlfarrell** [comment]
More on this.  if you try to edit the code during the idle reconnect refresh, the entire app breaks.  
![image](https://github.com/user-attachments/assets/1aed8427-2508-4b03-a226-6eb8d60adc59)

**Comment by mlfarrell** [comment]
Note that this also happened to me (long reconnect and flicker) _while_ being inside the app.  i.e. something seems very broken still.   keep this in mind with your new PR.

**Comment by lf94** [comment]
Oh! The case of disabling the code editor (and literally all the buttons) was completely forgotten. I'll do it in a follow up PR - thanks for pointing it out!

**Comment by lf94** [comment]
I'm unassigning because this is a long term issue that involves engine changes and we have other mitigations for now

**Comment by Irev-Dev** [comment]
Related: https://github.com/KittyCAD/modeling-app/issues/3020

**Comment by franknoirot** [comment]
@lee-at-zoo-corp I do not know if this is resolved by #5312. Much of this issue is stale;
- your settings are not lost on engine disconnect an reconnect
- all modeling toolbar buttons and commands are disabled when the engine connection is off
- while KCL errors telling you that the engine failed to respond occur if you type while your WiFi is off, for example, the errors disappear on reconnection

I am actually of the mind that this should be closed now, what do you think?

**Comment by lee-at-zoo-corp** [comment]
Let's wait until it's merged so we have more points to compare against - these are also good for writing tests against

**Comment by franknoirot** [comment]
@lee-at-zoo-corp is this resolved for you when you enable the new stream idle feature?

**Comment by lee-at-zoo-corp** [comment]
Yep, closing. _Camera_ state is now restored. We never idle on sketch mode so state is never lost there.
